rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+
+State transitions
+~~~~~~~~~~~~~~~~~
+
+The state transitions occur either as a result of an API call from the Client, or when the Bootloader carries out an installation operation. The installation operations that occur when the Bootloader runs are determined by the state of the Firmware Store, and do not depend on the reason for the restart. To ensure a specific action occurs at restart, the Client uses the appropriate API to set the state of the Firmware Store.
+
+The Client can trigger transitions in the state model using the following operations:
+
+* ``start``
+* ``write``
+* ``install``
+* ``erase``
+* ``accept``
+* ``reject``
+
+Some transitions can only be triggered by a system restart, which will run the Bootloader. For these transitions it does not matter how the reboot was triggered.
+
+The typical flow through the states is shown in this graphic:
+
+.. image:: fwu-states-simple.svg
+
+The depicted flow does not show the behavior in error scenarios, except for the transitions over reboot where a failure can only be reported to the Client by changing the state of the Firmware Store.
+
+The READY state at the end is different to the one at the start --- the *active* firmware image is the updated version. The Firmware Store is ready to start the process again from the beginning for the next update.
+
+State/operation transition matrix
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. list-table::
+    :header-rows: 2
+    :stub-columns: 1
+
+    * -
+      - Operation
+      -
+      -
+      -
+      -
+      -
+      -
+    * - State
+      - ``start``
+      - ``write``
+      - ``install``
+      - ``reboot``
+      - ``accept``
+      - ``reject``
+      - ``erase``
+
+    * - READY
+      - Begin update →WRITING
+      - *Error*
+      - *Error*
+      - no effect
+      - *Error*
+      - *Error*
+      - no effect
+    * - WRITING
+      - *Error*
+      - Write to *second*
+      - Verify *second* →STAGED
+      - no effect
+      - *Error*
+      - *Error*
+      - Clean *second* →READY
+    * - STAGED
+      - *Error*
+      - *Error*
+      - *Error*
+      - if install successful:
+          Swap images →TRIAL
+        else:
+          Record error →FAILED
+      - *Error*
+      - *Error*
+      - Clean *second* →READY
+    * - FAILED
+      - *Error*
+      - *Error*
+      - *Error*
+      - no effect
+      - *Error*
+      - *Error*
+      - Clean *second* →READY
+    * - TRIAL
+      - *Error*
+      - *Error*
+      - *Error*
+      - Swap images →FAILED
+      - Lock in update →UPDATED
+      - Reject update →REJECTED
+      - *Error*
+    * - REJECTED
+      - *Error*
+      - *Error*
+      - *Error*
+      - Swap images →FAILED
+      - *Error*
+      - *Error*
+      - *Error*
+    * - UPDATED
+      - *Error*
+      - *Error*
+      - *Error*
+      - no effect
+      - *Error*
+      - *Error*
+      - Clean *second* →READY
```

Comment:
I like the "clean" more than "erase" as it doesn't suggest it actually has to go to flash and erase the whole image.

Reply:
I have debated combining abort/erase/clean with start, as this part of the flow has no choice after `erase`. Perhaps calling this `prepare`? This would mean that if there is a long running flash operation to clean the image, it is deferred until a new image is ready to install. Would this be a problem?

If not, I guess `clean` works as an operation name, as it suggests cleaning the state, not just the image data. `erase` is almost too prescriptive, for situations where actually erasing the flash is not technically needed.